lib/bb/parse/parse_py/BBHandler.py

```diff
@@ -1,378 +1,260 @@
+
+    if ext != ".bbclass" and oldfile and abs_fn != oldfile:
+        d.setVar("FILE", oldfile)
+
     return d
 
-def feeder(lineno, s, fn, d):
-    global __func_start_regexp__, __inherit_regexp__, __export_func_regexp__, __addtask_regexp__, __addhandler_regexp__, __def_regexp__, __python_func_regexp__, __inpython__,__infunc__, __body__, __bbpath_found__, classes, bb, __residue__
+def feeder(lineno, s, fn, root, statements, eof=False):
+    global __func_start_regexp__, __inherit_regexp__, __export_func_regexp__, __addtask_regexp__, __addhandler_regexp__, __def_regexp__, __python_func_regexp__, __inpython__, __infunc__, __body__, bb, __residue__, __classname__
+
+    # Check tabs in python functions:
+    # - def py_funcname(): covered by __inpython__
+    # - python(): covered by '__anonymous' == __infunc__[0]
+    # - python funcname(): covered by __infunc__[3]
+    if __inpython__ or (__infunc__ and ('__anonymous' == __infunc__[0] or __infunc__[3])):
+        tab = __python_tab_regexp__.match(s)
+        if tab:
+            bb.warn('python should use 4 spaces indentation, but found tabs in %s, line %s' % (root, lineno))
+
     if __infunc__:
         if s == '}':
             __body__.append('')
-            data.setVar(__infunc__, '\n'.join(__body__), d)
-            data.setVarFlag(__infunc__, "func", 1, d)
-            if __infunc__ == "__anonymous":
-                anonqueue = bb.data.getVar("__anonqueue", d) or []
-                anonitem = {}
-                anonitem["content"] = bb.data.getVar("__anonymous", d)
-                anonitem["flags"] = bb.data.getVarFlags("__anonymous", d)
-                anonqueue.append(anonitem)
-                bb.data.setVar("__anonqueue", anonqueue, d)
-                bb.data.delVarFlags("__anonymous", d)
-                bb.data.delVar("__anonymous", d)
-            __infunc__ = ""
+            ast.handleMethod(statements, fn, lineno, __infunc__[0], __body__, __infunc__[3], __infunc__[4])
+            __infunc__ = []
             __body__ = []
         else:
             __body__.append(s)
         return
 
     if __inpython__:
         m = __python_func_regexp__.match(s)
-        if m:
+        if m and not eof:
             __body__.append(s)
             return
         else:
-            text = '\n'.join(__body__)
-            comp = compile(text, "<bb>", "exec")
-            exec comp in __builtins__
+            ast.handlePythonMethod(statements, fn, lineno, __inpython__,
+                                   root, __body__)
             __body__ = []
             __inpython__ = False
-            funcs = data.getVar('__functions__', d) or ""
-            data.setVar('__functions__', "%s\n%s" % (funcs, text), d)
-#           fall through
 
-    if s == '' or s[0] == '#': return          # skip comments and empty lines
+            if eof:
+                return
+
+    if s and s[0] == '#':
+        if len(__residue__) != 0 and __residue__[0][0] != "#":
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0` or `if len(val) &gt; 0`

 [Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

Similar issue at line number 183.